library/src/main/java/com/datatorrent/lib/stream/MultipleStreamMerger.java

```diff
@@ -0,0 +1,277 @@
+      throw new IllegalArgumentException("Not enough streams to merge, at least two streams must be selected for " +
+          "merging with `.merge()`.");
+    }
+
+    ArrayList<Stream> streamsToAddToDag = new ArrayList<>();
+    ArrayList<NamedMerger> operatorsToAdd = new ArrayList<>();
+
+    // Determine operators and streams to add to the DAG
+    constructMergeTree(streamsToAddToDag, operatorsToAdd);
+
+    for (NamedMerger m : operatorsToAdd) {
+      dag.addOperator(m.name, m.merger);
+    }
+
+    for (Stream s : streamsToAddToDag) {
+      dag.addStream(s.name, s.sourcePort, s.destPort).setLocality(DAG.Locality.CONTAINER_LOCAL);
+    }
+  }
+
+  /**
+   * Given a set of streams to be merged (defined via {@link #merge(DefaultOutputPort)}), compute the optimal
+   * structure of cascading mergers that need to be instantiated, added to the dag, and linked together.
+   * @param streamsToAddToDag - (output)  A list that is populated with streams that should be added to the  DAG
+   * @param operatorsToAdd - (output) A list that is populated with operators to be added to the DAG
+   */
+  public void constructMergeTree(
+      ArrayList<Stream> streamsToAddToDag,
+      ArrayList<NamedMerger> operatorsToAdd)
+  {
+    if (streamsToMerge.size() < 2) {
+      throw new IllegalArgumentException("Not enough streams to merge. Ensure `.merge` was called for each stream " +
+          "to be added.");
+    }
+
+    // Define the final merger in the sequence and connect its output to the module's output
+    StreamMerger<K> finalMerger = new StreamMerger<>();
+    operatorsToAdd.add(new NamedMerger("Merger_Final", finalMerger));
+    streamOutput.set(finalMerger.out);
+
+    ArrayList<ArrayList<StreamMerger<K>>> mergers = new ArrayList<>();
+
+    /**
+     * First, calculate the number of tiers we need to merge all streams given that each merger can only merge two
+     * streams at a time.
+     */
+    int numTiers = (int)Math.ceil(Math.log(streamsToMerge.size()) / Math.log(2));
+
+    // Handle the simple case where we only have a single tier (only two streams to merge)
+    if (numTiers == 1) {
+      assert (streamsToMerge.size() == 2);
+      streamsToAddToDag.add(new Stream("FinalMerge_Stream_0", streamsToMerge.get(0), finalMerger.data1));
+      streamsToAddToDag.add(new Stream("FinalMerge_Stream_1", streamsToMerge.get(1), finalMerger.data2));
+
+      // We don't need to add any operators since we've already added the final merger
+    } else {
+      Iterator<DefaultOutputPort<K>> streams = streamsToMerge.iterator();
+
+      // When assigning streams, we will switch between ports 1 and 2 as we use successive mergers.
+      boolean usePort1;
+
+      // For each tier, create the mergers in that tier, and connect the relevant streams
+      for (int i = 0; i < numTiers - 1; i++) {
+        int streamIdx = 0;
+        usePort1 = true;
+
+        int numMergers = (int)Math.ceil(streamsToMerge.size() / Math.pow(2, i + 1));
```

Comment:
Should this be floor instead of ceil? The one odd stream can be directly fed to the final tier.